### PR TITLE
Unpin and update database for SQLA 2.0

### LINKED ISSUE #6647: TST: sqlalchemy.exc.InvalidRequestError DatabaseEntry has been deleted
### Describe the bug

```
_______________________________________________________________________________ test_clear_database ________________________________________________________________________________

filled_database = <Table length=1>
 id  observation_time_start observation_time_end ... starred download_time size
str1          str3   ...-------- ... ------- ------------- ----
   4                    N/A                  N/A ...      No           N/A  N/A

    def test_clear_database(filled_database):
        assert len(filled_database) == 10
        filled_database.clear()
        assert not filled_database
        assert filled_database.session.query(JSONDump).all() == []
        assert filled_database.session.query(FitsHeaderEntry).all() == []
        assert filled_database.session.query(FitsKeyComment).all() == []
        assert filled_database.session.query(Tag).all() == []
>       filled_database.undo()

.tox/py310-sunpy/lib/python3.10/site-packages/sunpy/database/tests/test_database.py:731: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.tox/py310-sunpy/lib/python3.10/site-packages/sunpy/database/database.py:1087: in undo
    self._command_manager.undo(n)
.tox/py310-sunpy/lib/python3.10/site-packages/sunpy/database/commands.py:359: in undo
    command.undo()
.tox/py310-sunpy/lib/python3.10/site-packages/sunpy/database/commands.py:98: in undo
    operation.undo()
.tox/py310-sunpy/lib/python3.10/site-packages/sunpy/database/commands.py:161: in undo
    self.session.add(self.entry)
.tox/py310-sunpy/lib/python3.10/site-packages/sqlalchemy/orm/scoping.py:361: in add
    return self._proxied.add(instance, _warn=_warn)
.tox/py310-sunpy/lib/python3.10/site-packages/sqlalchemy/orm/session.py:3137: in add
    self._save_or_update_state(state)
.tox/py310-sunpy/lib/python3.10/site-packages/sqlalchemy/orm/session.py:3167: in _save_or_update_state
    self._save_or_update_impl(st_)
.tox/py310-sunpy/lib/python3.10/site-packages/sqlalchemy/orm/session.py:3804: in _save_or_update_impl
    self._update_impl(state)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <sqlalchemy.orm.session.Session object at 0x7feda710ee60>, state = <sqlalchemy.orm.state.InstanceState object at 0x7feda7736800>, revert_deletion = False

    def _update_impl(
        self, state: InstanceState[Any], revert_deletion: bool = False
    ) -> None:
        if state.key is None:
            raise sa_exc.InvalidRequestError(
                "Instance '%s' is not persisted" % state_str(state)
            )
    
        if state._deleted:
            if revert_deletion:
                if not state._attached:
                    return
                del state._deleted
            else:
>               raise sa_exc.InvalidRequestError(
                    "Instance '%s' has been deleted.  "
                    "Use the make_transient() "
                    "function to send this object back "
                    "to the transient state." % state_str(state)
                )
E               sqlalchemy.exc.InvalidRequestError: Instance '<DatabaseEntry at 0x7feda8cdc580>' has been deleted.  Use the make_transient() function to send this object back to the transient state.

.tox/py310-sunpy/lib/python3.10/site-packages/sqlalchemy/orm/session.py:3774: InvalidRequestError
```

### To Reproduce

Run sunpy test over at https://github.com/astropy/astropy-integration-testing/actions

Example log: https://github.com/astropy/astropy-integration-testing/actions/runs/3623315366/jobs/6109043995

### Screenshots

_No response_

### System Details

_No response_

### Installation method

pip

### PR DESCRIPTION
**Pull request by alasdairwilson**
fixes #6647

This PR contains the necessary changes for SQLAlchemy >=2.0

The only substantive change was to the undo method on CompositeOperation which were broken for a complete mystery to me, see my comment on a previous pull for the explanation for that change:

> So the failed test was a massive pain to track down the exact root cause. `Database.clear()` generates a bunch of commands on a `CompositeCommand` which are ordered by the type of thing removed (first tags from entries then tags then entries) for some reason adding them back in via undo, in this same order, worked in <2.0 and now not in 2.0. In 2.0 it results in tags being added to id's that do not exist, creating entries with those id's which are then invalid when you try to write the `DatabaseEntry` with the same id back into the database. The whole `make_transient()` error is a red herring in my opinion because actually the sunpy code was correctly setting the` _sa_instance_state.deleted` flag to False before restoring these entries.
> 
> I tried for a long time to figure out the cause rooted in the difference between 2.0 and <2.0 SQLa for an elegant solution but I could not but the basic solution is to either sort the list of operations in the Composite operation but I found this, in any non manual ordering, to normally result in the SQL database violating unique constrains as id's are automatically assigned prior to the correct DBentry being readded to the table. The alternative, which I have done at is to specifically search the list of _operations in a `CompositeOperation.undo` for `RemoveEntry `where `RemoveEntry `is a `DatabaseEntry `and do that First and then undo all other RemoveEntry's (e.g. Tag) and finally undo all other operations (e.g. `RemoveTag`).
> 
> I actually dunno the best way to push to this branch Can either merge it at your repo or replace this pull with my branch:
> 
> https://github.com/alasdairwilson/sunpy/tree/sql
> 
> Tests pass:
> 
> https://github.com/alasdairwilson/sunpy/actions/runs/4044537889/jobs/6954812107

### CODE REVIEW

**Review by nabobalis — APPROVED** [review]

**Comment by nabobalis** [comment]
Thank you!

I will want to see if the oldest deps passes as well, before we merge.

**Comment by nabobalis** [comment]
Oldest deps is passing:

Changelog bot is confused but the PR has the correct changelog.
Figure tests are a known failure.

Will check the online build, but I think we don't run most of them for the database anymore.

File: sunpy/database/commands.py [hunk 1]
```diff
@@ -94,8 +96,20 @@ def __call__(self):
...
+                # we also filter out ~database.commands.RemoveEntry where type is not ~database.tables.DatabaseEntry to do later.
+                if isinstance(operation.entry, DatabaseEntry):
+                    operation.undo()
         for operation in self._operations:
-            operation.undo()
+            if isinstance(operation, RemoveEntry):
+                if not isinstance(operation.entry, DatabaseEntry):
+                    # undo all ~database.commands.RemoveEntry where type is not ~database.tables.DatabaseEntry
+                    operation.undo()
+            elif not isinstance(operation, RemoveEntry):
```
Comment (ConorMacBride) [inline sunpy/database/commands.py:110]: This should be equivalent, although maybe it's more readable as is?
```suggestion
            else:
```

**Review by ConorMacBride — APPROVED** [review]

**Comment by nabobalis** [comment]
Online fails, I can't see anything database related at a glance.

**Comment by nabobalis** [comment]
I will backport to 4.0 now.

**Pull request merged**